File: youtube/lat2-loan/try2.py
import pandas as pd
import numpy as np
import warnings
import os
from sklearn.model_selection import train_test_split
from datetime import date, timedelta
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import roc_curve
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sympy import false

# print(os.getcwd())
# print(os.listdir())
warnings.filterwarnings('ignore')

loan_data = pd.read_csv('ML/youtube/lat2-loan/lc_2016_2017.csv')
print(loan_data.head())
print(loan_data.shape)
print(loan_data.info())
print(loan_data.isnull().sum())

print(loan_data.loan_status.unique())
loan_data['good_bad'] = np.where(
    loan_data['loan_status'].isin(['Charged Off', 'Default', 'Late (16-30 days)', 'Late (31-120 days)']),
    1,0
)
print(loan_data.head())
print(loan_data.good_bad.value_counts())
print(loan_data.good_bad.value_counts(normalize = True))

# null_data = pd.DataFrame(loan_data.isnull().sum() / loan_data.shape[0] *100)
null_data = pd.DataFrame(loan_data.isnull().sum() / len(loan_data) *100)
null_data = null_data[null_data.iloc[:,0] > 50].sort_values(0, ascending=False)
print(null_data)

# loan_data = loan_data.dropna(thresh=len(loan_data) * 0.5, axis = 1)
# loan_data = loan_data.dropna(thresh=loan_data.shape[0] * 0.5, axis = 1)
# loan_data = loan_data.dropna(thresh=len(loan_data) /2, axis = 1)
loan_data = loan_data.dropna(thresh=loan_data.shape[0] /2, axis = 1)

null_data = pd.DataFrame(loan_data.isnull().sum() / loan_data.shape[0] * 100, columns=['nullpct'])
null_data = null_data[null_data['nullpct'] > 50].sort_values('nullpct', ascending=False)
# null_data = null_data[null_data.loc[:, 'nullpct'] > 50].sort_values('nullpct', ascending=False)
print(null_data)

x = loan_data.drop('good_bad', axis = 1)
y = loan_data['good_bad']

X_train, X_test, y_train, y_test = train_test_split(x,y, test_size=0.2, random_state=42, stratify=y)
print(y_train.value_counts(normalize = True))
print(y_test.value_counts(normalize = True))
print(X_train.dtypes)

# for col in X_train.select_dtypes(include = ['object', 'bool']):
#     print(col)
#     print(X_train[col])
#     print()

for col in X_train.select_dtypes(include = ['object', 'bool']).columns:
    print(col)
    print(X_train[col])
    print()
    
col_need_to_clean = ['term', 'emp_length', 'issue_d', 'earliest_cr_line', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d']
print(X_train[col_need_to_clean].head())

X_train['term'] = pd.to_numeric(X_train['term'].str.replace(' months', ''))

print(X_train['emp_length'].unique())
X_train['emp_length'] = X_train['emp_length'].str.replace('+ years', '')
X_train['emp_length'] = X_train['emp_length'].str.replace('< 1 year', '0')
X_train['emp_length'] = X_train['emp_length'].str.replace(' years', '')
X_train['emp_length'] = X_train['emp_length'].str.replace(' year', '')
X_train['emp_length'] = X_train['emp_length'].fillna(0)
X_train['emp_length'] = pd.to_numeric(X_train['emp_length'])


print(X_train[col_need_to_clean].head())

penanggalan = ['issue_d', 'earliest_cr_line', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d']
for tanggal in penanggalan:
    X_train[tanggal] = pd.to_datetime(X_train[tanggal])

print(X_train[penanggalan].head())
print(X_train[col_need_to_clean].head())

print('\n\n\n\n\n\n')
print(X_test[col_need_to_clean].head())

X_test['term'] = pd.to_numeric(X_test['term'].str.replace(' months', ''))

print(X_test['emp_length'].unique())
X_test['emp_length'] = X_test['emp_length'].str.replace('+ years', '')
X_test['emp_length'] = X_test['emp_length'].str.replace('< 1 year', '0')
X_test['emp_length'] = X_test['emp_length'].str.replace(' years', '')
X_test['emp_length'] = X_test['emp_length'].str.replace(' year', '')
X_test['emp_length'] = X_test['emp_length'].fillna(0)
X_test['emp_length'] = pd.to_numeric(X_test['emp_length'])

for tanggal in penanggalan:
    X_test[tanggal] = pd.to_datetime(X_test[tanggal])
    
print(X_train[penanggalan].head())
print(X_train[col_need_to_clean].head())
print(X_test[penanggalan].head())
print(X_test[col_need_to_clean].head())
print(X_train[col_need_to_clean].info())
print(X_test[col_need_to_clean].info())

print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)
print(col_need_to_clean)

X_train = X_train[col_need_to_clean]
del X_train['next_pymnt_d']
X_test = X_test[col_need_to_clean]
X_test = X_test.drop('next_pymnt_d', axis = 1)

print(X_train.head())
print(X_test.head())
print(X_train.shape)
print(X_test.shape)
print(X_train.isnull().sum())
print(X_test.isnull().sum())
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)
print()

print(date.today().strftime('%Y-%m-%d'))
def date_columns(df, col):
    today = pd.to_datetime(date.today().strftime('%Y-%m-%d'))
    df[col] = pd.to_datetime(df[col], format = '%b-%y')
    df['mth since-' + col] = round(pd.to_numeric((today - df[col]) / np.timedelta64(30, 'D')))
    df.drop(col,axis = 1, inplace = True)
    
date_columns(X_train, 'issue_d') 
date_columns(X_train, 'earliest_cr_line')
date_columns(X_train, 'last_pymnt_d')
date_columns(X_train, 'last_credit_pull_d')

date_columns(X_test, 'issue_d')
date_columns(X_test, 'earliest_cr_line')
date_columns(X_test, 'last_pymnt_d')
date_columns(X_test, 'last_credit_pull_d')

X_train = X_train.fillna(X_train.median())
X_test = X_test.fillna(X_test.median())

print(X_train.head())
print(X_test.head())
print(X_train.shape)
print(X_test.shape)
print(X_train.isnull().sum())
print(X_test.isnull().sum())


k_folds = KFold(n_splits=5, shuffle=True, random_state=0)
model  = LogisticRegression()
scoring = 'accuracy'
score = cross_val_score(model, X_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))

model.fit(X_train, y_train)
y_pred = model.predict(X_test)
print(y_pred)



result = pd.DataFrame({
    'y_pred': y_pred,
    'y_test': y_test
})
print(result)

accuracy = accuracy_score(y_test, y_pred)
print('akurasi: {}'.format(accuracy))

cm = confusion_matrix(y_test, y_pred)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred')
plt.ylabel('y_test')
plt.title('confusion matrix')
plt.show()

y_pred = model.predict_proba(X_test)
print(y_pred)
# y_pred = model.predict_proba(X_test)[:,0]
# print('Yang Bagus')
# print(y_pred)
# print(y_pred > 0.5)
# print((y_pred > 0.5).astype(int))
# print((y_pred > 0.066).astype(int)) 
# print((y_pred < 0.066).astype(int))
y_pred = model.predict_proba(X_test)[:,1]
print('Yang Jelek')
print(y_pred)
print(y_pred > 0.5)
print((y_pred > 0.5).astype(int))
print((y_pred > 0.066).astype(int))
print((y_pred > 0.22).astype(int))
# print((y_pred < 0.066).astype(int))

plt.hist(y_pred)
plt.show()


print()
print()
fpr,tpr,thresholds = roc_curve(y_test, y_pred)
print(fpr)
print(tpr)
print(thresholds)
print()

# youden j-statistic
j = tpr - fpr
print(j)
ix = np.argmax(j)
print(ix)
best_thresh = thresholds[ix]
print(best_thresh)

y_pred = model.predict_proba(X_test)[:,1]
y_pred = (y_pred > 0.066).astype(int)
print('y_pred')
print(y_pred)
result = pd.DataFrame({
    'y_pred': y_pred,
    'y_test': y_test
})
print(result)
accuracy = accuracy_score(y_test, y_pred)
print('akurasi: {}'.format(accuracy))
cm = confusion_matrix(y_test, y_pred)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred')
plt.ylabel('y_test')
plt.title('confusion matrix')
plt.show()

print(model.coef_)
print(model.intercept_)

df_coeff = pd.DataFrame(model.coef_)
print(df_coeff)
df_coeff = pd.DataFrame(model.coef_, columns=X_train.columns)
print(df_coeff)

File: youtube/lat2-loan/try2.r
library(dplyr)
library(ggplot2)
library(stringr)
library(gridExtra)
library(RMySQL)
library(dplyr)
library(zoo)
library(pROC)
library(ROCR)
library(tidyr)
library(plyr)
library(lubridate)
library(randomForest)
library(caret)
library(patchwork)
library(nnet)
library(klaR) # untuk Naive Bayes
library(rpart) # untuk Decision Trees


loan_data <- read.csv("ML/youtube/lat2-loan/lc_2016_2017.csv", na.strings = c("", "NA", " ", "N/A", "NULL"))
print(head(loan_data))
print(dim(loan_data))
print(colSums(is.na(loan_data)))

# loan_data$good_bad <- case_when(
#     loan_data$loan_status %in% c('Charged Off', 'Default', 'Late (16-30 days)', 'Late (31-120 days)') ~ 1,
#     .default = 0
# )

loan_data$good_bad <- ifelse(
    loan_data$loan_status %in% c("Charged Off", "Default", "Late (16-30 days)", "Late (31-120 days)"), 1, 0
)

print(head(loan_data))
print(table(loan_data$good_bad))
print(prop.table(table(loan_data$good_bad)))

# null_data <- data.frame(null = colSums(is.na(loan_data)) / dim(loan_data)[1] * 100)
# null_data <- data.frame(null = colSums(is.na(loan_data)) / nrow(loan_data) * 100)
null_data <- data.frame(null = colMeans(is.na(loan_data)) * 100)
null_data <- null_data %>%
    filter(null > 50) %>%
    arrange(desc(null))
print(null_data)

loan_data <- loan_data[, colMeans(is.na(loan_data)) < 0.5]

# null_data <- data.frame(null = colSums(is.na(loan_data)) / dim(loan_data)[1] * 100)
null_data <- data.frame(null = colSums(is.na(loan_data)) / nrow(loan_data) * 100)
# null_data <- data.frame(null = colMeans(is.na(loan_data)) * 100)
null_data <- null_data %>%
    filter(null > 50) %>%
    arrange(desc(null))
print(null_data)

x <- dplyr::select(loan_data, -good_bad)
y <- as.factor(loan_data$good_bad)
set.seed(42)
index <- createDataPartition(y, p = 0.8, list = TRUE)

x_train <- x[index$Resample1, ]
x_test <- x[-index$Resample1, ]
y_train <- y[index$Resample1]
y_test <- y[-index$Resample1]

print(prop.table(table(y_train)))
print(prop.table(table(y_test)))

for (col in colnames(x_train[, sapply(x_train, function(x) is.character(x) | is.logical(x))])) {
    print(col)
    print(head(x_train[[col]]))
    cat("\n\n\n")
}

col_need_to_clean <- c("term", "emp_length", "issue_d", "earliest_cr_line", "last_pymnt_d", "next_pymnt_d", "last_credit_pull_d")
x_train$term <- as.numeric(str_replace_all(x_train$term, " months", ""))

x_train$emp_length <- str_replace_all(x_train$emp_length, "\\+ years", "")
x_train$emp_length <- str_replace_all(x_train$emp_length, "< 1 year", "0")
x_train$emp_length <- str_replace_all(x_train$emp_length, " years", "")
x_train$emp_length <- str_replace_all(x_train$emp_length, " year", "")
x_train$emp_length[is.na(x_train$emp_length)] <- 0
x_train$emp_length <- as.numeric(x_train$emp_length)

penanggalan <- c("issue_d", "earliest_cr_line", "last_pymnt_d", "next_pymnt_d", "last_credit_pull_d")
for (tanggal in penanggalan) {
    x_train[[tanggal]] <- my(x_train[[tanggal]])
}

print(head(x_train[col_need_to_clean]))



x_test$term <- as.numeric(str_replace_all(x_test$term, " months", ""))

x_test$emp_length <- str_replace_all(x_test$emp_length, "\\+ years", "")
x_test$emp_length <- str_replace_all(x_test$emp_length, "< 1 year", "0")
x_test$emp_length <- str_replace_all(x_test$emp_length, " years", "")
x_test$emp_length <- str_replace_all(x_test$emp_length, " year", "")
x_test$emp_length[is.na(x_test$emp_length)] <- 0
x_test$emp_length <- as.numeric(x_test$emp_length)


for (tanggal in penanggalan) {
    x_test[[tanggal]] <- my(x_test[[tanggal]])
}

print(head(x_test[col_need_to_clean]))


x_train <- x_train[col_need_to_clean]
x_train <- dplyr::select(x_train, -next_pymnt_d)
x_test <- x_test[col_need_to_clean]
x_test <- subset(x_test, select = -next_pymnt_d)

print(head(x_train))
print(head(x_test))

date_columns <- function(df, col) {
    today <- ymd(today())
    df[[col]] <- ymd(df[[col]])
    df[paste("mth since-", col)] <- round(as.numeric((today - df[[col]]) / 30))
    df <- dplyr::select(df, -col)
    # df <- subset(df, select = -col)
    return(df)
}

x_train <- date_columns(x_train, "issue_d")
x_train <- date_columns(x_train, "earliest_cr_line")
x_train <- date_columns(x_train, "last_pymnt_d")
x_train <- date_columns(x_train, "last_credit_pull_d")

x_test <- date_columns(x_test, "issue_d")
x_test <- date_columns(x_test, "earliest_cr_line")
x_test <- date_columns(x_test, "last_pymnt_d")
x_test <- date_columns(x_test, "last_credit_pull_d")

x_train <- na.aggregate(x_train, FUN = median) 
x_test <- na.aggregate(x_test, FUN = median) 

print(head(x_train))
print(head(x_test))
print(dim(x_train))
print(dim(x_test))
print(colSums(is.na(x_train)))
print(colSums(is.na(x_test)))



ctrl <- trainControl(method = "cv", number = 5)
# modelrf <- train(x_train, y_train, method = "rf", trControl = ctrl)
# print("accuracy mean")
# print(modelrf$result)
# print(mean(modelrf$result$Accuracy))
# cat("\n")

# modelnnet <- train(x_train, y_train, method="nnet", trControl=ctrl)
# print('accuracy mean')
# print(modelnnet$result)
# print(mean(modelnnet$result$Accuracy))
# cat('\n')

modelrpart <- train(x_train, y_train, method="rpart", trControl=ctrl)
print('accuracy mean')
print(modelrpart$result)
print(mean(modelrpart$result$Accuracy))
cat('\n')

# modelnb <- train(x_train, y_train, method="nb", trControl=ctrl)
# print('accuracy mean')
# print(modelnb$result)
# print(mean(modelnb$result$Accuracy))
# cat('\n')

y_pred <- predict(modelrpart, x_test)
print(head(y_pred))


df_pred <- data.frame(
    y_test = y_test,
    y_pred = y_pred
)
print(head(df_pred))

hasil <- df_pred[df_pred$y_test != df_pred$y_pred, ]
# hasil <- filter(df_pred, y_test != y_pred)
print(head(hasil))

accuracy <- confusionMatrix(data = y_pred, reference = y_test)$overall["Accuracy"]
print(accuracy)


y_pred <- predict(modelrpart, x_test, type = 'prob')
print(head(y_pred))


print('bagus')
y_pred <- predict(modelrpart, x_test, type = "prob")[,1]
print(head(y_pred))
print(head(y_pred > 0.5))
print(head(as.integer(y_pred > 0.5)))
print('jelek')
y_pred <- predict(modelrpart, x_test, type = "prob")[,2]
print(head(y_pred))
print(head(y_pred > 0.5))
print(head(as.integer(y_pred > 0.5)))


# Menghitung kurva ROC
roc_data <- roc(y_test, y_pred)

# Mendapatkan nilai FPR, TPR, dan thresholds
fpr <- roc_data$specificities
tpr <- roc_data$sensitivities
thresholds <- roc_data$thresholds

# Menampilkan hasil
print("FPR:")
print(fpr)
print("\nTPR:")
print(tpr)
print("\nThresholds:")
print(thresholds)

# Menghitung Youden's J statistic
j <- tpr - fpr
best_ix <- which.max(j)
best_thresh <- thresholds[best_ix]


cat("\n\nYouden's J Statistic:")
print(j)
cat("\nIndex of Best Threshold:")
print(best_ix)
cat("\nBest Threshold:")
print(best_thresh)

y_pred <- predict(modelrpart, x_test, type = "prob")[,2]
y_pred <- as.factor(as.numeric(y_pred > 0.066))

df_pred <- data.frame(
    y_test = y_test,
    y_pred = y_pred
)
print(head(df_pred))

hasil <- df_pred[df_pred$y_test != df_pred$y_pred, ]
# hasil <- filter(df_pred, y_test != y_pred)
print(head(hasil))

accuracy <- confusionMatrix(data = y_pred, reference = y_test)$overall["Accuracy"]
print(accuracy)